flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecWindowAggregateBase.java

```diff
@@ -0,0 +1,114 @@
+import org.apache.flink.table.planner.plan.logical.WindowSpec;
+import org.apache.flink.table.planner.plan.logical.WindowingStrategy;
+import org.apache.flink.table.planner.plan.nodes.exec.ExecNode;
+import org.apache.flink.table.planner.plan.nodes.exec.InputProperty;
+import org.apache.flink.table.runtime.operators.window.slicing.SliceAssigner;
+import org.apache.flink.table.runtime.operators.window.slicing.SliceAssigners;
+import org.apache.flink.table.types.logical.LogicalType;
+
+import java.time.Duration;
+import java.util.List;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/** The base class for window aggregate {@link ExecNode}. */
+public abstract class StreamExecWindowAggregateBase extends StreamExecAggregateBase {
+
+    protected StreamExecWindowAggregateBase(
+            int id,
+            List<InputProperty> inputProperties,
+            LogicalType outputType,
+            String description) {
+        super(id, inputProperties, outputType, description);
+        checkArgument(inputProperties.size() == 1);
+    }
+
+    // ------------------------------------------------------------------------------------------
+    // Utilities
+    // ------------------------------------------------------------------------------------------
+
+    protected SliceAssigner createSliceAssigner(WindowingStrategy windowingStrategy) {
+        WindowSpec windowSpec = windowingStrategy.getWindow();
+        if (windowingStrategy instanceof WindowAttachedWindowingStrategy) {
+            int windowEndIndex =
+                    ((WindowAttachedWindowingStrategy) windowingStrategy).getWindowEnd();
+            // we don't need time attribute to assign windows, use a magic value in this case
+            SliceAssigner innerAssigner = createSliceAssigner(windowSpec, Integer.MAX_VALUE);
+            return SliceAssigners.windowed(windowEndIndex, innerAssigner);
+
+        } else if (windowingStrategy instanceof TimeAttributeWindowingStrategy) {
+            final int timeAttributeIndex;
+            if (windowingStrategy.isRowtime()) {
+                timeAttributeIndex =
+                        ((TimeAttributeWindowingStrategy) windowingStrategy)
+                                .getTimeAttributeIndex();
+            } else {
+                timeAttributeIndex = -1;
+            }
+            return createSliceAssigner(windowSpec, timeAttributeIndex);
+
+        } else {
+            throw new UnsupportedOperationException(windowingStrategy + " is not supported yet.");
+        }
+    }
+
+    protected SliceAssigner createSliceAssigner(WindowSpec windowSpec, int timeAttributeIndex) {
+        if (windowSpec instanceof TumblingWindowSpec) {
+            Duration size = ((TumblingWindowSpec) windowSpec).getSize();
+            return SliceAssigners.tumbling(timeAttributeIndex, size);
+
+        } else if (windowSpec instanceof HoppingWindowSpec) {
```

Comment:
Maybe we could move those check codes to `isValid` method in `StreamPhysicalWindowAggregate`?
BTW, Currently, some validate code exists in  `isValid` method of `StreamPhysical**Node`, others exists in `StreamExec**Node. I'm a littler confused, is there any rule to apply when we decides where to validate node?

Reply:
The problem is there will be duplicate codes. But I think we can improve it in the future.